File: lib/hubclient/hubconnect/ConnectToHub.go
package hubconnect

import (
	"fmt"
	"github.com/hiveot/hub/lib/certs"
	"github.com/hiveot/hub/lib/discovery"
	"github.com/hiveot/hub/lib/hubclient"
	"github.com/hiveot/hub/lib/hubclient/mqtthubclient"
	"github.com/hiveot/hub/lib/hubclient/natshubclient"
	"log/slog"
	"path"
	"strings"
	"time"
)

// ConnectToHub helper function to connect to the Hub using existing token and key files.
// This assumes that CA cert, user keys and auth token have already been set up.
//
// The format for the key file is {clientID}.nkey for nats and {clientID}Key.pem for mqtt.
// For format for the token file is {clientID}.token.
//
// 1. If no fullURL is given then use discovery to determine the URL
// 2. Determine the core to use
// 3. Load the CA cert
// 4. Create a hub client
// 5. Connect using token and key files
func ConnectToHub(fullURL string, clientID string, certDir string, core string) (
	hc hubclient.IHubClient, err error) {

	var keyFile string
	var tokenFile string

	// 1. determine the actual address
	if fullURL == "" {
		// return after first result
		fullURL, core = discovery.LocateHub(time.Second, true)
	}
	if clientID == "" {
		return nil, fmt.Errorf("missing clientID")
	}
	// 2. obtain the CA public cert to verify the server
	caCertFile := path.Join(certDir, certs.DefaultCaCertFile)
	caCert, err := certs.LoadX509CertFromPEM(caCertFile)
	if err != nil {
		return nil, err
	}
	// 3. Determine which core to use and setup the key and token filenames
	// By convention the key/token filename format is "{name}.key/{name}.token"
	tokenFile = path.Join(certDir, clientID+".token")
	keyFile = path.Join(certDir, clientID+".key")
	if core == "nats" || strings.HasPrefix(fullURL, "nats") {
		// nats with nkeys. The key filename format is "{serviceID}.key"
		hc = natshubclient.NewNatsHubClient(fullURL, clientID, nil, caCert)
	} else {
		// mqtt with ecdsa keys. The key filename format is "{serviceID}.key|token"
		hc = mqtthubclient.NewMqttHubClient(fullURL, clientID, nil, caCert)
	}
	// 4. Connect and auth with token
	slog.Info("connecting to", "serverURL", fullURL)
	err = hc.ConnectWithTokenFile(tokenFile, keyFile)
	if err != nil {
		return nil, err
	}
	return hc, err
}

File: libjs/src/hubclient/ConnectToHub.ts
import {IHubClient} from "@hivelib/hubclient/IHubClient";
import {locateHub} from "@hivelib/hubclient/locateHub";
import path from "path";
import {HttpSSEClient} from "@hivelib/hubclient/httpclient/HttpSSEClient";



// ConnectToHub helper function to connect to the Hub using existing token and key files.
// This assumes that CA cert, user keys and auth token have already been set up and
// are available in the certDir.
// The key-pair file is named {certDir}/{clientID}.key
// The token file is named {certDir}/{clientID}.token
//
// 1. If no fullURL is given then use discovery to determine the URL
// 2. Determine the core to use
// 3. Load the CA cert
// 4. Create a hub client
// 5. Connect using token and key files
//
//	fullURL is the scheme://addr:port/[wspath] the server is listening on
//	clientID to connect as. Also used for the key and token file names
//	certDir is the location of the CA cert and key/token files
// This throws an error if a connection cannot be made
export async function ConnectToHub(
    fullURL: string, clientID: string, caCertPem: string): Promise<IHubClient> {

    // 1. determine the actual address
    if (fullURL == "") {
        // return after first result
        let uc = await locateHub()
        fullURL = uc.hubURL
    }
    if (clientID == ""||fullURL == "") {
        throw("Missing clientID or hub URL")
    }
    // 2. Determine the client protocol to use
    // TODO: support multiple client protocols
    let hc = new HttpSSEClient(fullURL, clientID, caCertPem)
    return hc
}
